### PR TITLE
feat: add metrics to repo overview

### LINKED ISSUE #222: Consider some "top level" stats/metrics to display in `/repos` page?
Above the main repos table, maybe some things like:

- Total number of repos (that match the current filters - tags and search terms)
- Avg. number of syncs per repo
- Avg. sync run time

etc. Basically some numbers (maybe small charts?) to give the user an aggregate sense of what their repo and sync data looks like

### PR DESCRIPTION
**Pull request by emilyvandewalle**
- Started adding top metrics
- I was able to figure out the `Number of repos` and `Avg number of syncs`, but not `Avg sync time` for which I could use some help

image

(Resolves #222)

### CODE REVIEW

**Comment by patrickdevivo** [comment]
This looks great! @amenowanna and I discussed a bit and we wanted to change the numbers a bit. Implementation may need to involve @german-mergestat too, so tagging him in. But overall look and feel wise looks great.

The numbers should be:
- Total number of repos (stays the same)
- Total number of repo syncs (i.e. the sum of all the "sum" values in the table)
- Total number of repo syncs where the latest run had an error (i.e. the sum of all the "error" values in the table)

The other ask is to see if we can move the repo search bar to _below_ these numbers, to make it clear that they are global totals (and that searching does not change them). Thanks so much @emilyvandewalle

**Comment by amenowanna** [comment]
Looks great! My only comment here is that the titles of each metric is using different wording. Can we make it consistent please. Like `Total number of...` or `Number of...` @patrickdevivo have a preference?

**Comment by emilyvandewalle** [comment]
Oops, good catch, agree this should be consistent, adjusted it to `Total number of...` ! 👍

**Comment by emilyvandewalle** [comment]
(I was a bit too fast in trying to adjust it, I'll wait for @patrickdevivo to give his preference)

**Comment by patrickdevivo** [comment]
(Copied from Slack)

- Let’s change the text to just say Total repos Total repo syncs and Latest syncs with errors
- Let’s add a loading/indeterminate state for  the “metric box” when the data is fetching from the API (you can trigger this by writing a search query - tho these should be unrelated, I think we should actually pull them out into separate queries, especially since pagination will change that list-repos query)
- Finally the last box (Latest syncs with errors) needs to be implemented data wise

I think once those are addressed it should be good to merge!

**Comment by patrickdevivo** [comment]
```graphql
query {
  repos {
    totalCount
  }
  repoSyncs(condition: {scheduleEnabled: true}) {
    totalCount
  }
  execSQL(query: "WITH ranked_completed_syncs AS ( SELECT rsq.id, rsq.repo_sync_id, rsq.done_at, RANK() OVER(PARTITION BY rsq.repo_sync_id ORDER BY rsq.done_at DESC) AS rank_num FROM mergestat.repo_sync_queue AS rsq WHERE rsq.repo_sync_id NOT IN (SELECT repo_sync_id FROM mergestat.repo_sync_queue WHERE status = 'RUNNING' OR status = 'QUEUED') ) SELECT COUNT(DISTINCT rcs.id) AS syncs_error_count FROM ranked_completed_syncs AS rcs INNER JOIN mergestat.repo_sync_logs AS rsl ON rcs.id = rsl.repo_sync_queue_id WHERE rank_num = 1 AND log_type = 'ERROR'")
}
```

**Comment by patrickdevivo** [comment]
image

File: ui/src/views/repositories/index.tsx [hunk 1]
```diff
@@ -2,43 +2,113 @@ import { useRepositoriesContext } from 'src/state/contexts/repositories.context'
...
+                  <MetricNumber loading={loading} metric={metrics?.totalRepoSyncs || 0} />
+                </Stat.Number>
+              </Stat.Left>
+              <Stat.Right>
+                <ColoredBox size='12'><TableIcon className='t-icon t-icon-default' /></ColoredBox>
+              </Stat.Right>
+            </Stat>
+            <Stat className='shadow-sm w-full'>
+              <Stat.Left>
+                <Stat.Label>Latest syncs with errors</Stat.Label>
```
Comment (patrickdevivo) [inline ui/src/views/repositories/index.tsx:88]:
```suggestion
                <Stat.Label>Total syncs with errors</Stat.Label>
```

**Review by patrickdevivo — CHANGES_REQUESTED** [review]

File: ui/src/views/repositories/index.tsx (hunk 1 above)
Comment (german-mergestat) [reply ui/src/views/repositories/index.tsx]: Done [4d2715a](https://github.com/mergestat/mergestat/pull/414/commits/4d2715a2d33d47e1f4d444176bcbe34426c762e9)

**Review by patrickdevivo — APPROVED** [review]

**Pull request merged**